File: abgui/Sources/abgui/Models/JSONValue.swift
// Copyright 2026 Gigaion, LLC
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

/// A resilient decoder for Apple's open JSON:API attribute bags: whatever fields Apple
/// adds still decode, and per-screen typed structs pull the columns a view needs. Keeps
/// abgui from breaking when Apple grows an `attributes` object.
enum JSONValue: Codable, Equatable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
        } else if let b = try? c.decode(Bool.self) {
            self = .bool(b)
        } else if let n = try? c.decode(Double.self) {
            self = .number(n)
        } else if let s = try? c.decode(String.self) {
            self = .string(s)
        } else if let a = try? c.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? c.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(in: c, debugDescription: "unrecognized JSON value")
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let n): try c.encode(n)
        case .bool(let b):   try c.encode(b)
        case .object(let o): try c.encode(o)
        case .array(let a):  try c.encode(a)
        case .null:          try c.encodeNil()
        }
    }

    /// The string at `key` if this is an object with a string there — the common lookup
    /// for a table column (name, type, serialNumber, …).
    func string(_ key: String) -> String? {
        if case .object(let o) = self, case .string(let s)? = o[key] { return s }
        return nil
    }

    /// The array at `key` if this is an object with an array there (e.g. a user's `roles`).
    func array(_ key: String) -> [JSONValue]? {
        if case .object(let o) = self, case .array(let a)? = o[key] { return a }
        return nil
    }
}

File: abgui-flutter/lib/src/models/json_value.dart
// Copyright 2026 Gigaion, LLC
// SPDX-License-Identifier: AGPL-3.0-or-later

// Port note (rule 5): the Swift `JSONValue` is a six-case Codable enum
// (string/number/bool/object/array/null). Here it is a thin wrapper around the already-decoded
// `Object?` that `jsonDecode` hands back, because that is what keeps the CALL SITES simple —
// and the call sites are the whole reason the type exists.
//
// Nothing in abgui ever pattern-matches all six cases: the views ask "the string at this key"
// (`Resource.attr`) or "the array at this key" (`Resource.roleNames`), which is exactly the
// pair of accessors the Swift enum grew. A sealed hierarchy would add six classes and a
// `switch` at every use to answer the same two questions, and re-wrapping every nested value
// on decode would cost a full copy of every attribute bag for no reader that wants it.
//
// The behaviour that matters is preserved exactly: a lookup whose value is not a string
// answers null (see `string`), so a boolean `isCustomApp` still renders as the table's em dash
// instead of the text "false", and unknown keys survive a round trip untouched.

import 'json.dart';

/// A resilient container for Apple's open JSON:API attribute bags: whatever fields Apple adds
/// still decode, and per-screen accessors pull the columns a view needs. Keeps abgui from
/// breaking when Apple grows an `attributes` object.
class JSONValue {
  /// The decoded JSON, verbatim: a `Map`, `List`, `String`, `num`, `bool`, or null.
  final Object? raw;

  const JSONValue(this.raw);

  /// Decoding is the identity function — see the port note above.
  factory JSONValue.fromJson(Object? value) => JSONValue(value);

  /// True when the value is JSON `null` (distinct from "this key is absent", which the
  /// containing model represents with a null `JSONValue?`).
  bool get isNull => raw == null;

  /// The string at [key] if this is an object with a string there — the common lookup for a
  /// table column (name, type, serialNumber, …).
  String? string(String key) {
    final object = raw;
    if (object is Map) {
      final value = object[key];
      if (value is String) return value;
    }
    return null;
  }

  /// The array at [key] if this is an object with an array there (e.g. a user's `roles`).
  List<JSONValue>? array(String key) {
    final object = raw;
    if (object is Map) {
      final value = object[key];
      if (value is List) {
        return value.map(JSONValue.new).toList(growable: false);
      }
    }
    return null;
  }

  /// The object at [key] wrapped for further lookups, or null.
  JSONValue? object(String key) {
    final object = raw;
    if (object is Map) {
      final value = object[key];
      if (value is Map) return JSONValue(value);
    }
    return null;
  }

  /// Re-encodable as-is; the wrapper never rewrites what it was handed.
  Object? toJson() => raw;

  @override
  bool operator ==(Object other) =>
      other is JSONValue && deepJsonEquals(raw, other.raw);

  @override
  int get hashCode => deepJsonHash(raw);

  @override
  String toString() => 'JSONValue($raw)';
}
